bentoml/_internal/runner/runner.py

```diff
@@ -183,238 +140,179 @@ def get_runnable_methods(cls) -> t.List[RunnableMethod]:
-        if self._runner_app_client:
-            return self._runner_app_client.run(runner_method_name, *args, **kwargs)
+    def init_client(self, runner: Runner):
+        if self._runner_client:
+            logger.warning("re creating remote runner client")
         if self._runnable:
-            return self._runnable[runner_method_name](*args, **kwargs)
+            raise BentoMLException("TODO: ..")
 
+        self._runner_client = RemoteRunnerClient(runner)
+
+    def destroy_client(self):
+        if not self._runner_client:
+            logger.warning("remote runner client not found")
+        else:
+            del self._runner_client
+
+    def run_method(
+        self,
+        method_name: str,
+        *args: t.Any,
+        **kwargs: t.Any,
+    ) -> t.Any:
+        if self._runnable is not None:
+            return getattr(self._runnable, method_name)(*args, **kwargs)
+        if self._runner_client is not None:
+            return self._runner_client.run_method(method_name, *args, **kwargs)
         raise BentoMLException(
             "runner not initialized"
         )  # TODO: make this UninitializedRunnerException
 
-    async def _async_run(self, runner_method_name, *args, **kwargs):
-        if self._runner_app_client:
-            return await self._runner_app_client.async_run(
-                runner_method_name, *args, **kwargs
-            )
-        if self._runnable:
+    async def async_run_method(
+        self,
+        method_name: str,
+        *args: t.Any,
+        **kwargs: t.Any,
+    ) -> t.Any:
+        if self._runnable is not None:
             import anyio
 
-            # TODO(jiang): to_thread
-            # return await self._runnable[runner_method_name](*args, **kwargs)
-
+            method = getattr(self._runnable, method_name)
+            return anyio.to_thread.run_sync(method, *args, **kwargs)
+        if self._runner_client is not None:
+            return await self._runner_client.async_run_method(
+                method_name,
+                *args,
+                **kwargs,
+            )
         raise BentoMLException(
             "runner not initialized"
         )  # TODO: make this UninitializedRunnerException
 
+
+@attr.define(frozen=True)
+class Runner:
+    runnable_class: t.Type[Runnable] = attr.field()
+    init_params: t.Dict[str, t.Any] = attr.field()
+    name: str = attr.field()
+    strategy: t.Type[Strategy] = attr.field(default=DefaultStrategy)
+    models: t.List[Model] = attr.field(factory=list)
+    resource_config: Resource = attr.field()
+    method_configs = attr.field()
+    runner_handle = attr.field(init=False)
+
+    def __init__(
+        self,
+        runnable_class: t.Type[Runnable],
+        init_params: t.Dict[str, t.Any],
+        name: str,
+        strategy: t.Type[Strategy],
+        models: t.List[Model],
+        cpu: int,
+        nvidia_gpu: int,
+        custom_resources: t.Dict[str, int | float],
```

Comment:
Would it be cleaner if we ask users to pass a `Resource` instance. The `Resource` instance can be overridden to add custom resources as properties. The custom `Strategy` will know how to consume the custom `Resource)

Reply:
It means user always need to import an extra class for passing in resource config, I feel passing in the parameters directly is nicer from user perspective. Resource class is nice for internal use and validation. We took a similar approach in Model and BuildConfig as well

Reply:
Passing `Resource` may give better extensibility in the future. Say when we add `another_type_gpu`, we won't need to update the signature here. We can also get rid of the `custom_resources` field because a custom implementation can add any additional fields that are only needed to be understand by their custom strategies.